File: backend_fastAPi/services/agent_service.py
import re
from typing import Dict, Any, Optional
from services.llm_service import LLMService, LLMServiceError
import json
import asyncio  # Add this import

class AgentService:
    """Service that implements AI agent behavior for UI generation."""
    
    def __init__(self):
        self.llm_service = LLMService()
        
    async def process_requirement(self, requirement: str) -> Dict[str, Any]:
        try:
            is_modification = self._is_modification_request(requirement)
            
            analysis = await self._analyze_requirement(requirement, is_modification)
            plan = await self._plan_implementation(requirement, analysis, is_modification)
            
            # Generate React project files instead of basic UI code
            generated_files = await self._generate_react_project_files(requirement, analysis, plan)
            
            return {
                "files": generated_files,
                "analysis": analysis,
                "plan": plan,
                "feedback": None
            }
            
        except Exception as e:
            raise LLMServiceError(f"Failed to process requirement: {str(e)}")
    
    def _is_modification_request(self, requirement: str) -> bool:
        """Check if the requirement is requesting a modification to an existing UI."""
        modification_keywords = [
            "add", "change", "modify", "update", "remove", "delete", 
            "reset button", "alter", "adjust", "extend", "enhance"
        ]
        
        requirement_lower = requirement.lower()
        return any(keyword in requirement_lower for keyword in modification_keywords)
    
    async def _analyze_requirement(self, requirement: str, is_modification: bool) -> str:
        """Analyze the requirement with React + shadcn/ui context."""
        prompt = f"""
        You are a React development expert. Analyze this UI requirement:
        
        '{requirement}'
        
        Consider:
        1. What React components are needed
        2. Which shadcn/ui components can be used
        3. What state management is required
        4. Component hierarchy and data flow
        5. Any technical constraints or challenges
        
        Return only your analysis, formatted clearly.
        """
        
        return await self.llm_service.generate_text(prompt)
    
    async def _plan_implementation(self, requirement: str, analysis: str, is_modification: bool) -> str:
        """Create an implementation plan for React + shadcn/ui."""
        prompt = f"""
        Based on:
        Requirement: '{requirement}'
        Analysis: '{analysis}'
        
        Create a step-by-step plan to implement this React + Vite + shadcn/ui project:
        1. Project structure and file organization
        2. Component breakdown and responsibilities
        3. State management approach
        4. Implementation order and dependencies
        5. Integration points between components
        
        Return only the concrete implementation plan, formatted as a clear list.
        """
        
        return await self.llm_service.generate_text(prompt)
    
    async def _generate_react_project_files(self, requirement: str, analysis: str, plan: str) -> Dict[str, str]:
        try:
            prompt = f"""
            Create a React + TypeScript project for this TODO app requirement:
            {requirement}

            Return ONLY a valid JSON object with these file paths (do not include any others):
            {{
                "/src/App.tsx": "<React component code>",
                "/src/components/TodoList.tsx": "<component code>",
                "/src/components/TodoItem.tsx": "<component code>"
            }}

            Rules:
            1. Use absolute paths starting with /src/
            2. Include only TypeScript/React code files
            3. Do not include config files, they will be added separately
            4. Return only valid JSON with the specified files
            5. Each component should be a separate file
            6. Use proper TypeScript types
            """

            response = await self.llm_service.generate_text(prompt)
            
            try:
                files = json.loads(response)
                # Ensure all paths start with /src/
                files = {
                    k if k.startswith("/src/") else f"/src/{k.lstrip('/')}": v
                    for k, v in files.items()
                }
                return files
            except json.JSONDecodeError:
                raise LLMServiceError("Failed to parse generated code as JSON")

        except Exception as e:
            raise LLMServiceError(f"React project generation failed: {str(e)}")
    
    async def _generate_feedback(self, requirement: str, analysis: str) -> str:
        """Generate feedback when the requirement is too vague or not feasible."""
        prompt = f"""
        The following UI requirement appears to be unclear or not feasible:
        
        '{requirement}'
        
        Based on this analysis: '{analysis}'
        
        Please provide:
        1. A clear explanation of what makes this requirement challenging
        2. Specific questions that would help clarify the requirement
        3. Alternative suggestions that might meet the user's needs
        
        Format this as helpful feedback to the user.
        """
        
        return await self.llm_service.generate_text(prompt)
    
    def _extract_code_blocks(self, text: str) -> tuple[Optional[str], Optional[str], Optional[str]]:
        """Extract HTML, CSS, and JavaScript code blocks from the generated text."""
        if not isinstance(text, str):
            raise ValueError("Generated code must be a string")

        def extract_block(pattern: str) -> Optional[str]:
            matches = re.search(pattern, text, re.DOTALL | re.IGNORECASE)
            return matches.group(1).strip() if matches else None
            
        html = extract_block(r"```html\s*(.*?)\s*```")
        css = extract_block(r"```css\s*(.*?)\s*```")
        javascript = extract_block(r"```javascript\s*(.*?)\s*```")
        
        if not any([html, css, javascript]):
            raise ValueError("No code blocks found in generated text")
            
        return html or "", css or "", javascript or ""

File: backend_node/services/agent-service.js
// Agent Service for UI generation
const { LLMService, LLMServiceError } = require("./llm-service");

class AgentService {
  constructor() {
    this.llmService = new LLMService();
  }

  // Check if the requirement is requesting a modification to an existing UI
  _isModificationRequest(requirement) {
    const modificationKeywords = [
      "add",
      "change",
      "modify",
      "update",
      "remove",
      "delete",
      "reset button",
      "alter",
      "adjust",
      "extend",
      "enhance",
    ];

    const requirementLower = requirement.toLowerCase();
    return modificationKeywords.some((keyword) =>
      requirementLower.includes(keyword)
    );
  }

  // Analyze the requirement with Next.js + Tailwind context
  async _analyzeRequirement(requirement, isModification) {
    const prompt = `
    You are a Next.js development expert. Analyze this UI requirement:
    
    '${requirement}'
    
    Consider:
    1. What Next.js components are needed
    2. What Tailwind CSS classes would be appropriate
    3. What state management is required
    4. Component hierarchy and data flow
    5. Any technical constraints or challenges
    
    Return only your analysis, formatted clearly.
    `;

    return await this.llmService.generateText(prompt);
  }

  // Create an implementation plan for Next.js + Tailwind
  async _planImplementation(requirement, analysis, isModification) {
    const prompt = `
    Based on:
    Requirement: '${requirement}'
    Analysis: '${analysis}'
    
    Create a step-by-step plan to implement this Next.js + Tailwind CSS project:
    1. Project structure and file organization
    2. Component breakdown and responsibilities (using plain JavaScript, not TypeScript)
    3. State management approach with React hooks
    4. Implementation order and dependencies
    5. Styling approach using Tailwind CSS classes
    
    Return only the concrete implementation plan, formatted as a clear list.
    `;

    return await this.llmService.generateText(prompt);
  }

  // Generate Next.js project files
  async _generateNextJsProjectFiles(requirement, analysis, plan) {
    try {
      const prompt = `
      🧠 Requirement:
      ${requirement}

      ✅ Tasks to perform:

      1. Create a Next.js project (no TypeScript).
      2. Setup Tailwind CSS with necessary configuration files.
      3. Create the following project structure and files:

      📁 File Structure:
      - /pages/index.js → main page rendering the TODO app
      - /components/TodoList.js → displays list of todo items
      - /components/TodoItem.js → represents individual todo item
      - /styles/globals.css → global styles using Tailwind
      - /tailwind.config.js → Tailwind configuration
      - /postcss.config.js → PostCSS configuration
      - /package.json → all dependencies listed (Next.js, React, TailwindCSS)

      🧩 Logic Requirements:
      - index.js should include state for managing todos (add/delete/toggle)
      - TodoList should accept todos as props (using plain JavaScript, not TypeScript)
      - TodoItem should accept todo, onToggle, and onDelete as props
      - Use ONLY Tailwind CSS classes for styling (no inline styles, no CSS modules)
      - Use plain HTML elements with Tailwind classes (no component libraries)
      - Components should be functional components written in plain modern JavaScript

      🎁 Output Format:
      Return a single valid JSON object with file paths as keys and code content as values.

      🚫 Important:
      - Do NOT include TypeScript extensions (.ts, .tsx)
      - Do NOT use TypeScript syntax (no interfaces, types, or React.FC)
      - Do NOT use shadcn/ui or any other UI component libraries
      - Do NOT include src/ in file paths
      - Do NOT include explanations or markdown - ONLY JSON

      ✅ Example Output:
      {
        "/pages/index.js": "import { useState } from 'react'...",
        "/components/TodoList.js": "export default function TodoList({ todos, onToggle, onDelete }) {...",
        ...
      }
      `;

      const response = await this.llmService.generateText(prompt);

      try {
        // Clean response by removing markdown formatting and extracting JSON
        const cleanResponse = response.replace(/```(json)?/g, "").trim();
        const jsonMatch = cleanResponse.match(/\{[\s\S]*\}/);

        if (!jsonMatch) {
          throw new Error("No valid JSON found in response");
        }

        let files;
        try {
          files = JSON.parse(jsonMatch[0]);
        } catch (jsonError) {
          // Try to sanitize the JSON if it has issues
          const sanitizedJson = jsonMatch[0]
            .replace(/\\'/g, "'") // Fix escaped single quotes
            .replace(/\\"/g, '"') // Fix escaped double quotes
            .replace(/\n/g, "\\n"); // Properly escape newlines

          files = JSON.parse(sanitizedJson);
        }

        // Normalize file paths to ensure they start with /
        const normalizedFiles = {};
        for (const [key, value] of Object.entries(files)) {
          // Ensure key starts with / and doesn't have /src/
          let normalizedKey = key.startsWith("/") ? key : `/${key}`;
          // Remove any /src/ prefix
          normalizedKey = normalizedKey.replace(/^\/src\//, "/");

          // Check for TypeScript extensions
          if (normalizedKey.endsWith(".tsx") || normalizedKey.endsWith(".ts")) {
            const jsPath = normalizedKey.replace(/\.tsx?$/, ".js");
            console.warn(
              `Converting TypeScript path to JavaScript: ${normalizedKey} → ${jsPath}`
            );
            normalizedKey = jsPath;
          }

          normalizedFiles[normalizedKey] = value;
        } // Validate required files
        const requiredFiles = [
          "/pages/index.js",
          "/components/TodoList.js",
          "/components/TodoItem.js",
          "/styles/globals.css",
          "/tailwind.config.js",
          "/postcss.config.js",
          "/package.json",
        ];

        const missingFiles = requiredFiles.filter(
          (file) => !normalizedFiles[file]
        );
        if (missingFiles.length > 0) {
          console.warn(
            `Warning: Missing required files: ${missingFiles.join(", ")}`
          );
        }

        // Validate file extensions to ensure they're .js not .ts/.tsx
        const tsFiles = Object.keys(normalizedFiles).filter(
          (file) => file.endsWith(".ts") || file.endsWith(".tsx")
        );

        if (tsFiles.length > 0) {
          console.warn(
            `Warning: TypeScript files detected: ${tsFiles.join(", ")}`
          );
        }

        return normalizedFiles;
      } catch (error) {
        if (error instanceof SyntaxError) {
          throw new Error(
            `Failed to parse generated code as JSON: ${error.message}`
          );
        }
        throw error;
      }
    } catch (error) {
      throw new Error(`Next.js project generation failed: ${error.message}`);
    }
  }

  // Generate feedback when the requirement is too vague or not feasible
  async _generateFeedback(requirement, analysis) {
    const prompt = `
    The following UI requirement appears to be unclear or not feasible:
    
    '${requirement}'
    
    Based on this analysis: '${analysis}'
    
    Please provide:
    1. A clear explanation of what makes this requirement challenging
    2. Specific questions that would help clarify the requirement
    3. Alternative suggestions that might meet the user's needs
    
    Format this as helpful feedback to the user.
    `;

    return await this.llmService.generateText(prompt);
  }

  // Process the requirement
  async processRequirement(requirement) {
    try {
      const isModification = this._isModificationRequest(requirement);

      const analysis = await this._analyzeRequirement(
        requirement,
        isModification
      );
      const plan = await this._planImplementation(
        requirement,
        analysis,
        isModification
      );

      // Generate Next.js project files instead of basic UI code
      const generatedFiles = await this._generateNextJsProjectFiles(
        requirement,
        analysis,
        plan
      );

      return {
        files: generatedFiles,
        analysis: analysis,
        plan: plan,
        feedback: null,
      };
    } catch (error) {
      throw new Error(`Failed to process requirement: ${error.message}`);
    }
  }
}

module.exports = { AgentService };
